thrift.js

```diff
@@ -121,33 +133,120 @@ function Thrift(options) {
-        source = this.fs.readFileSync(this.filename, 'ascii');
-        this.idls[this.filename] = source;
-    }
-
     this.exception = null;
 
-    // Separate compile/link passes permits forward references and cyclic
-    // references.
-    this.compile(source);
-    // We only link from the root Thrift object.
-    if (!options.noLink) {
-        this.link();
-    }
+    var model = this;
+    this._parse(this.filename, this.allowIncludeAlias, function (err) {
+        if (err) {
+            return model._exitOnError(loadedCb, err);
+        }
+
+        // Separate compile/link passes permits forward references and cyclic
+        // references.
+        try {
+            model.compile();
+            // We only link from the root Thrift object.
+            if (!options.noLink) {
+                model.link();
+            }
+        } catch (err) {
+            return model._exitOnError(loadedCb, err);
+        }
+        if (loadedCb) {
+            loadedCb(undefined, model);
+        }
+    });
 }
 
 Thrift.prototype.models = 'module';
 
 Thrift.prototype.Message = Message;
 
+Thrift.prototype._exitOnError = function _exitOnError(loadedCb, err) {
+    if (loadedCb) {
+        loadedCb(err, undefined);
+    } else {
+        throw err;
+    }
+}
+
+Thrift.prototype._syncFileLoader = function _syncFileLoader(fs) { 
+    return function(filename, cb) {
+        var source;
+        var error;
+        try {
+            /* eslint-disable max-len */
+            assert.ok(fs, filename + ': Thrift must be constructed with either a complete set of options.idls, options.asts, or options.fs access');
+            /* eslint-enable max-len */
+            source = fs.readFileSync(path.resolve(filename), 'ascii');
+        } catch (err) {
+            error = err;
+        }
+        cb(error, source);
+    };
+}
+
+Thrift.prototype._parse = function _parse(filename, allowIncludeAlias, cb) {
+    if (this.parsed[filename]) {
+        return cb(undefined);
+    }
+    this.parsed[filename] = true;
+
+    var model = this;
+    var loader = this.loader;
+    if (this.idls[filename] || this.asts[filename]) {
+        loader = function (_, cb) { cb(undefined, model.idls[filename]); };
+    }
+    loader(filename, function (err, source) {
+        if (err) {
+            return cb(err);
+        }
+        model.idls[filename] = source;
+        if (!model.asts[filename]) {
+            try {
+                model.asts[filename] = idl.parse(source);
+            } catch (err) {
+                return cb(err);
+            }
+        }
+        var defs = model.asts[filename].headers.concat(model.asts[filename].definitions);
+        model._parseIncludedModules(filename, defs, allowIncludeAlias, cb);
+    });
+}
+
+Thrift.prototype._parseIncludedModules = function _parseIncludedModules(filename, defs, allowIncludeAlias, cb) {
+    var model = this;
+    var dirname = path.dirname(filename);
+    async.eachOf(defs, function (def, _, callback) {
```

Comment:
If we can avoid a dependency on async, we should. It’s not much trouble to inline.

Reply:
I created a custom alternative to `async.eachOf`, see the tests PR for tests for it. It is not incredibly complicated, but still not trivial. I can refactor to use a simple recursive function loading sources one after the other instead of all at the same time if you prefer.